onnxruntime/core/providers/webnn/builders/impl/lstm_op_builder.cc

```diff
@@ -0,0 +1,199 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Copyright (c) Intel Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/common.h"
+#include "core/providers/shared/utils/utils.h"
+#include "core/providers/webnn/builders/helper.h"
+#include "core/providers/webnn/builders/model_builder.h"
+#include "core/providers/webnn/builders/op_builder_factory.h"
+
+#include "base_op_builder.h"
+
+namespace onnxruntime::webnn {
+
+class LstmOpBuilder : public BaseOpBuilder {
+  // Add operator related.
+ public:
+  void AddInitializersToSkip(ModelBuilder& model_builder, const Node& node) const override;
+
+ private:
+  Status AddToModelBuilderImpl(ModelBuilder& model_builder, const Node& node,
+                               const logging::Logger& logger) const override ORT_MUST_USE_RESULT;
+
+  // Operator support related.
+ private:
+  bool IsOpSupportedImpl(const InitializedTensorSet& initializers, const Node& node,
+                         const WebnnDeviceType /*device_type*/, const logging::Logger& logger) const override;
+};
+
+void LstmOpBuilder::AddInitializersToSkip(ModelBuilder& model_builder, const Node& node) const {
+  if (node.InputDefs().size() > 4 && node.InputDefs()[4]->Exists()) {
+    model_builder.AddInitializerToSkip(node.InputDefs()[4]->Name());  // sequence_lens
+    model_builder.AddInputToSkip(node.InputDefs()[4]->Name());
+  }
+}
+
+Status LstmOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder, const Node& node,
+                                            const logging::Logger& logger) const {
+  NodeAttrHelper helper(node);
+  uint32_t hidden_size = helper.Get("hidden_size", 1);
+
+  const auto& input_defs = node.InputDefs();
+  std::vector<int64_t> input_shape;
+  ORT_RETURN_IF_NOT(GetShape(*input_defs[0], input_shape, logger), "Cannot get input's shape");
+  uint32_t steps = static_cast<uint32_t>(input_shape[0]);
+
+  emscripten::val input = model_builder.GetOperand(input_defs[0]->Name());
+  emscripten::val weight = model_builder.GetOperand(input_defs[1]->Name());
+  emscripten::val recurrent_weight = model_builder.GetOperand(input_defs[2]->Name());
+
+  emscripten::val options = emscripten::val::object();
```

Comment:
Missing `label` option for each WebNN operators, you can use node name as context.